File: frontend/src/api/operations.ts
import { apiClient } from "./client";

export interface Kompi {
  id: number;
  nama: string;
  deskripsi?: string;
  created_at: string;
}

export interface Pleton {
  id: number;
  kompi_id: number;
  nama: string;
  deskripsi?: string;
  kompi?: Kompi;
  created_at: string;
}

export interface Regu {
  id: number;
  pleton_id: number;
  station_id?: number;
  nama: string;
  deskripsi?: string;
  pleton?: Pleton;
  created_at: string;
}

export interface Personil {
  id: number;
  regu_id?: number;
  nip_nik: string;
  nama_lengkap: string;
  jabatan: string;
  no_hp?: string;
  foto_url?: string;
  is_active: boolean;
  regu?: Regu;
  created_at: string;
}

// KOMPI API
export const getKompis = () => apiClient.get<Kompi[]>("/kompi").then((res: any) => res.data);
export const createKompi = (data: any) => apiClient.post<Kompi>("/kompi", data).then((res: any) => res.data);
export const updateKompi = (id: number, data: any) => apiClient.put<Kompi>(`/kompi/${id}`, data).then((res: any) => res.data);
export const deleteKompi = (id: number) => apiClient.delete(`/kompi/${id}`).then((res: any) => res.data);

// PLETON API
export const getPletons = () => apiClient.get<Pleton[]>("/pleton").then((res: any) => res.data);
export const createPleton = (data: any) => apiClient.post<Pleton>("/pleton", data).then((res: any) => res.data);
export const updatePleton = (id: number, data: any) => apiClient.put<Pleton>(`/pleton/${id}`, data).then((res: any) => res.data);
export const deletePleton = (id: number) => apiClient.delete(`/pleton/${id}`).then((res: any) => res.data);

// REGU API
export const getRegus = () => apiClient.get<Regu[]>("/regu").then((res: any) => res.data);
export const createRegu = (data: any) => apiClient.post<Regu>("/regu", data).then((res: any) => res.data);
export const updateRegu = (id: number, data: any) => apiClient.put<Regu>(`/regu/${id}`, data).then((res: any) => res.data);
export const deleteRegu = (id: number) => apiClient.delete(`/regu/${id}`).then((res: any) => res.data);

// PERSONIL API
export const getPersonils = () => apiClient.get<Personil[]>("/personil").then((res: any) => res.data);
export const createPersonil = (data: any) => apiClient.post<Personil>("/personil", data).then((res: any) => res.data);
export const updatePersonil = (id: number, data: any) => apiClient.put<Personil>(`/personil/${id}`, data).then((res: any) => res.data);
export const deletePersonil = (id: number) => apiClient.delete(`/personil/${id}`).then((res: any) => res.data);

File: backend/app/schemas/operations.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List, Any
from datetime import datetime
from app.models.personil import JabatanPersonil

class KompiBase(BaseModel):
    nama: str
    deskripsi: Optional[str] = None

class KompiCreate(KompiBase):
    pass

class KompiUpdate(BaseModel):
    nama: Optional[str] = None
    deskripsi: Optional[str] = None

class KompiResponse(KompiBase):
    id: int
    tenant_id: Any
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

class PletonBase(BaseModel):
    kompi_id: int
    nama: str
    deskripsi: Optional[str] = None

class PletonCreate(PletonBase):
    pass

class PletonUpdate(BaseModel):
    kompi_id: Optional[int] = None
    nama: Optional[str] = None
    deskripsi: Optional[str] = None

class PletonResponse(PletonBase):
    id: int
    tenant_id: Any
    kompi: Optional[KompiResponse] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

class ReguBase(BaseModel):
    pleton_id: int
    station_id: Optional[int] = None
    nama: str
    deskripsi: Optional[str] = None

class ReguCreate(ReguBase):
    pass

class ReguUpdate(BaseModel):
    pleton_id: Optional[int] = None
    station_id: Optional[int] = None
    nama: Optional[str] = None
    deskripsi: Optional[str] = None

class ReguResponse(ReguBase):
    id: int
    tenant_id: Any
    pleton: Optional[PletonResponse] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

class PersonilBase(BaseModel):
    user_id: Optional[int] = None
    regu_id: Optional[int] = None
    nip_nik: str
    nama_lengkap: str
    jabatan: JabatanPersonil
    no_hp: Optional[str] = None
    foto_url: Optional[str] = None
    is_active: bool = True

class PersonilCreate(PersonilBase):
    email: Optional[str] = None
    password: Optional[str] = None  # Will default to nip_nik if not provided

class PersonilUpdate(BaseModel):
    regu_id: Optional[int] = None
    nip_nik: Optional[str] = None
    nama_lengkap: Optional[str] = None
    jabatan: Optional[JabatanPersonil] = None
    no_hp: Optional[str] = None
    foto_url: Optional[str] = None
    is_active: Optional[bool] = None
    email: Optional[str] = None
    password: Optional[str] = None

class PersonilResponse(PersonilBase):
    id: int
    tenant_id: Any
    regu: Optional[ReguResponse] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)
